matrix_sdk/src/client.rs

```diff
@@ -646,6 +649,94 @@ impl Client {
         self.send(request).await
     }
 
+    /// Search the homeserver's directory of public rooms.
+    ///
+    /// Returns a `get_public_rooms::Response`, an empty response.
+    ///
+    /// # Arguments
+    ///
+    /// * `limit` - The number of `PublicRoomsChunk`s in each response.
+    ///
+    /// * `since` - Pagination token from a previous request.
+    ///
+    /// * `server` - The name of the server, if `None` the requested server is used.
+    ///
+    /// # Examples
+    /// ```no_run
+    /// use matrix_sdk::{Client, RoomSearchBuilder};
+    /// # use matrix_sdk::api::r0::directory::get_public_rooms;
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("http://example.com").unwrap();
+    /// # let limit = Some(10);
+    /// # let since = Some("since token");
+    /// # let server = Some("server name");
+    ///
+    /// let mut cli = Client::new(homeserver).unwrap();
+    /// # use futures::executor::block_on;
+    /// # block_on(async {
+    /// assert!(cli.get_public_rooms(
+    ///     limit,
+    ///     since,
+    ///     server
+    /// ).await.is_ok());
+    /// # });
+    /// ```
+    pub async fn get_public_rooms(
+        &self,
+        limit: Option<u32>,
+        since: Option<&str>,
+        server: Option<&str>,
+    ) -> Result<get_public_rooms::Response> {
+        let limit = limit.map(|n| UInt::try_from(n).ok()).flatten();
+        let since = since.map(ToString::to_string);
+        let server = server.map(ToString::to_string);
+
+        let request = get_public_rooms::Request {
+            limit,
+            since,
+            server,
+        };
+        self.send(request).await
+    }
+
+    /// Search the homeserver's directory of public rooms filtered.
+    ///
+    /// Returns a `get_public_rooms_filtered::Response`, an empty response.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_search` - The easiest way to create this request is using the `RoomSearchBuilder`.
+    ///
+    /// # Examples
+    /// ```
+    /// # use std::convert::TryFrom;
```

Comment:
Why are these lines prepended with hashes?

Reply:
Oh, it's a rust doc feature. Neat!

Reply:
To hide those uninteresting lines. Explained [here](https://doc.rust-lang.org/rustdoc/documentation-tests.html#hiding-portions-of-the-example).

Reply:
@dkasak I wanted to mention that I looked over everything you've been doing, everything looked great that was a nice PR! I like the renaming to `hoist_prev_content...` I could not think of a good one.

Reply:
Thanks! JS was the inspiration for that one, for better or worse. :smile: